tools/qfvis/source/flow.c

```diff
@@ -300,6 +300,234 @@ free_winding_memory (threaddata_t *thread, size_t winding_mark)
 	Hunk_RawFreeToLowMark (thread->hunk, winding_mark);
 }
 
+/*
+	TargetChecks
+
+	Filter mightsee by clipping against all portals
+*/
+static unsigned
+TargetChecks(threaddata_t *thread,
+			 pstack_t *prevstack,
+			 set_t **mainportalbits,
+			 set_t **tempportalbits)
+{
+	unsigned    i, initial_portalcheck;
+	portal_t   *target_portal;
+	vec4f_t     backplane;
+	vec4f_t     source_plane, pass_plane;
+	const winding_t *pass_winding;
+	winding_t  *source_winding, *target_winding;
+
+	if (prevstack->pass_winding == NULL)
+		return 0;
+	
+	thread->stats.chains++;
+
+	// initialize counter
+	initial_portalcheck = thread->stats.portalcheck;
+
+	// reuse separators
+	sep_t *separators[2] = {NULL, NULL};
+	if (prevstack->next) {
+		separators[0] = prevstack->next->separators[0];
+		separators[1] = prevstack->next->separators[1];
+	}
+
+	// empty cluster set
+	set_t *mightsee = thread->scratchbits;
+	set_empty(mightsee);
+
+	// empty portal set
+	set_t *prevportalbits = *mainportalbits;
+	set_t *portalbits = *tempportalbits;
+	set_empty(portalbits);
+	*mainportalbits = portalbits;
+	*tempportalbits = prevportalbits;
+
+	source_plane = thread->pstack_head.pass_plane;
+	pass_winding = prevstack->pass_winding;
+	pass_plane = prevstack->pass_plane;
+
+	// lower than 2 won't have a benefit. higher than 2 refines the source
+	// portal even though it isn't propagated forward right now.
+	int level = 2; // = options.level
+
+	size_t      winding_mark = Hunk_LowMark (thread->hunk);
+
+	// check all portals for flowing into other leafs
+	for (i = 0, target_portal = portals; i < numportals * 2; i++, target_portal++) {
+
+		if (!set_is_member(prevstack->mightsee, target_portal->cluster))
+            continue;           // can't possibly see it
+		if (!set_is_member(prevportalbits, i))
+            continue;           // can't possibly see it
+		if (set_is_member(mightsee, target_portal->cluster)) {
+			// target check already done and passed
+			set_add(portalbits, i); // let the portal survive to the next level
+            continue;           
+		}
+
+		// get plane of target_portal, point normal into the neighbor cluster
+		backplane = -target_portal->plane;
+
+		vec4f_t     diff = vabs4f (pass_plane - backplane);
+		vec4i_t     cmp = diff > (vec4f_t) {0.001, 0.001, 0.001, 0.001};
+		if (!(cmp[0] || cmp[1] || cmp[2])) { // dist isn't interesting
+			continue;		// can't go out a coplanar face
+		}
+
+		free_winding_memory (thread, winding_mark);
+
+		thread->stats.portalcheck++;
+
+		target_winding = target_portal->winding;
+		target_winding = ClipWinding (thread, target_winding, source_plane, false);
+		if (!target_winding)
+			continue;
+
+		if (!pass_winding) {
+			// unreachable
+		}
+
+		target_winding = ClipWinding (thread, target_winding, pass_plane, false);
+		if (!target_winding)
+			continue;
+
+		// copy source_winding because it likely is already a copy and thus
+		// if it gets clipped away, earlier stack levels will get corrupted
```

Comment:
A better comment might be "source_winding is owned by earlier stack levels, so copy it to avoid corruption"